packages/flutter/lib/src/services/text_input.dart

```diff
@@ -1184,123 +1363,33 @@ class TextInput {
   static TextInputConnection attach(TextInputClient client, TextInputConfiguration configuration) {
     assert(client != null);
     assert(configuration != null);
-    final TextInputConnection connection = TextInputConnection._(client);
-    _instance._attach(connection, configuration);
+    final TextInputConnection connection = _instance._connectionFactory(client);
+    connection.setClient(configuration);
+    _instance._currentConnection = connection;
     return connection;
   }
 
-  /// This method actually notifies the embedding of the client. It is utilized
-  /// by [attach] and by [_handleTextInputInvocation] for the
-  /// `TextInputClient.requestExistingInputState` method.
-  void _attach(TextInputConnection connection, TextInputConfiguration configuration) {
-    assert(connection != null);
-    assert(connection._client != null);
-    assert(configuration != null);
-    assert(_debugEnsureInputActionWorksOnPlatform(configuration.inputAction));
-    _channel.invokeMethod<void>(
-      'TextInput.setClient',
-      <dynamic>[ connection._id, configuration.toJson() ],
-    );
-    _currentConnection = connection;
-    _currentConfiguration = configuration;
+  /// Registers a factory method for creating [TextInputConnection] instances.
+  ///
+  /// By default, [TextInputConnection] communicates with the platform's text
+  /// input plugin via [SystemChannels.textInput].
+  static void registerConnectionFactory(TextInputConnectionFactory? factory) {
```

Comment:
What would become of the currently connected client if a different factory is registered? Should we make this thing an `InheritedWidget` so the current client (that has an open connection) can switch to the new factory, or we shouldn't support that use case at all(i.e., this should only be called before `runApp` and should never be called again)?

Reply:
It's fine to register and unregister a factory on the fly, but it's a factory so it will only affect the next client/connection.

Reply:
> nit: can this be static void registerConnectionFactory(TextInputConnectionFactory factory = _defaultConnectionFactory) {?

I wasn't able to resolve this request, because `TextInputConnection` (it increments a running client ID) doesn't have a const constructor, so I couldn't make `registerConnectionFactory` const either.

Reply:
> It's fine to register and unregister a factory on the fly, but it's a factory so it will only affect the next client/connection.

This means that if there's an active `TextField` and the factory is changed on the fly, then the user won't be able to use the new text input device unless they switch to a different field and switch back right? IMO we should either document that the developer shouldn't change the factory on the fly unless they can guarantee there's no active connection, or provide a means to notify the current client that the factory has been changed and it needs to reestablish the connection.

Reply:
I'm feeling a bit unsure if it's a good idea to make it fully automatic, but I have added a check for an active connection in `setInputConnectionFactory()`. It hides the old text input control and re-attaches the same client so that a new connection type is established.